core/src/identity.rs

```diff
@@ -116,6 +116,33 @@ impl Keypair {
         }
     }
 
+    /// Encode a private key as protobuf structure.
+    pub fn to_protobuf_encoding(&self) -> Result<Vec<u8>, DecodingError> {
```

Comment:
Can you expand on your use-case for encoding private keys?

Reply:
Here is a usecase: https://github.com/comit-network/rendezvous-server/blob/c5dedb5119569ff523da981f5bb3b02d3e3cb903/src/main.rs#L143-L151

We want the server to always have the same identity, which is why we store the secret key in a file. Using normal unix user ACLs, we can control which parts of the machine have access to this private key.

Our design was inspired by: https://smallstep.com/blog/command-line-secrets/

Reply:
Same case here: a bootstrap server (set of bootstrap servers) with a fixed identity.

Reply:
It might be a bit paranoid, but I would really like to keep the extra safety measure of _not being able to extract private key material once moved to a `Keypar`_.

Why not do something along the lines of [`GenerateNodeKeyCmd`](https://github.com/paritytech/substrate/blob/0ae6aac3b0f8bea55bf91737a179fa01cd50f13f/client/cli/src/commands/generate_node_key.rs#L40-L57) and [`InspectNodeKeyCmd`](https://github.com/paritytech/substrate/blob/0ae6aac3b0f8bea55bf91737a179fa01cd50f13f/client/cli/src/commands/inspect_node_key.rs#L41-L56)?

I would even suggest that we alter our [peer ID generator](https://github.com/libp2p/rust-libp2p/blob/master/misc/peer-id-generator/src/main.rs), enabling it to generate a random private key and saving it to a file, preferably in the IPFS format. The rendezvous-server or your bootstrap server can then read that file and load it as a `Keypair`.

I am taking a similar approach with [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/) and [ipfs init](https://github.com/ipfs/go-ipfs).

Reply:
Makes sense. I was confused that `Keypair::generate_*()` was a thing, with seemingly no way to persist the private key. Maybe this is a documentation issue after all, if it's indeed possible to construct the private information, persist it, and then yield a key.

Would you agree if I say that it's strange to generate a Ed25519 keypair, persist the secret (somehow) and then read that as a Keypair? I wouldn't expect a struct to serialize, then deserialize to a different type.

It's not too paranoid IMO.

> I would even suggest that we alter our peer ID generator, enabling it to generate a random private key and saving it to a file, preferably in the IPFS format. The rendezvous-server or your bootstrap server can then read that file and load it as a Keypair.

Note that saving to files is not a general solution; I imagine some people want to persist to a database, either in a file or in memory, etc.

Reply:
> Would you agree if I say that it's strange to generate a Ed25519 keypair, persist the secret (somehow) and then read that as a Keypair? I wouldn't expect a struct to serialize, then deserialize to a different type.

See https://github.com/libp2p/rust-libp2p/issues/2091#issuecomment-854540997 on the serialization format as a valid interface.

> Note that saving to files is not a general solution; I imagine some people want to persist to a database, either in a file or in memory, etc.

For sure. Any blob-store would work.

> Maybe this is a documentation issue after all

A patch to the docs would be very much appreciated, so others don't get confused too. Also, if you want to extend the peer ID generator, I would happily accept an upstream patch :innocent:

Reply:
> A patch to the docs would be very much appreciated, so others don't get confused too.

Making a lot of promises against `rust-libp2p` lately, but since I'll have to revise my own stuff now (it's currently using this branch), I suppose it's not a lot of work to create some docs and example code with doctests on the `Keypair` struct.

Reply:
> It might be a bit paranoid, but I would really like to keep the extra safety measure of _not being able to extract private key material once moved to a `Keypar`_.

But that is already possible or am I missing something?

- Fields in enums are public, hence one can just match on the variant and get for example an `ed25519::Keypair`
- `ed25519::Keypair` can be converted `.into()` an `ed25519::SecretKey`.
- `ed25519::SecretKey` implements `AsRef<[u8]>`

Reply:
Having a look at how to apply the approach you take in `GenerateNodeKeyCmd` and `InspectNodeKeyCmd`, but those seem not to use the protobuf encoding (which is what IPFS seems to use).

If the PeerID generator is altered to allow writing IPFS-style secret keys, I don't currently see how it can do without a patch like this one. Am I missing something here?

Reply:
:man_facepalming: my bad, you are absolutely right @thomaseizinger! Sorry for the confusion.

> If the PeerID generator is altered to allow writing IPFS-style secret keys, I don't currently see how it can do without a patch like this one. Am I missing something here?

Again, 100% correct.

Reply:
No worries, better to triple check and make sure! Thanks for having this :-)